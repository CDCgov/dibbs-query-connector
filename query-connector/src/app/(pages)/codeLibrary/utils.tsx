/**
 * @param system string
 * @returns string
 */
export const formatSystem = (system: string) => {
  switch (true) {
    case system.includes("snomed"):
      return "SNOMED";
    case system.includes("rxnorm"):
      return "RXNorm";
    case system.includes("http://hl7.org/fhir/sid/icd-10-cm"):
      return "ICD-10";
    default:
      return "";
  }
};
