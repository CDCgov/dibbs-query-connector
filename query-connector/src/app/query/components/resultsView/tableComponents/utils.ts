/**
 * Helper function to not display tables in results view where there are no
 * elements in a column to display. Allows for non-lengthwise properties
 * @param array - an array of items to display in a table
 * @param propertiesToCheck - the property that we want to conditionally display
 * in a table column if some element exists
 * @returns true or false for whether there is any element that exists for the
 * column to be rendered
 */
export function checkIfSomeElementWithPropertyExists<
  T extends object,
  K extends keyof T,
>(array: T[], propertiesToCheck: K[]): Record<K, boolean> {
  const result: Record<K, boolean> = propertiesToCheck.reduce(
    (acc, p) => {
      acc[p] = false;
      return acc;
    },
    {} as Record<K, boolean>,
  );

  for (const e of array) {
    for (const p of propertiesToCheck) {
      const prop = e[p];
      if ((prop as string) in e) {
        result[p] = true;
      }
    }
  }
  console.log(result);

  return result;
}
