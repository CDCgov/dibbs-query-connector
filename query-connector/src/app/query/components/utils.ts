import { GroupedValueSet } from "./customizeQuery/customizeQueryUtils";

/**
 *
 */
export const countDibbsConceptTypeToVsMapItems = (obj: {
  [vsNameAuthorSystem: string]: GroupedValueSet;
}) => {
  return Object.values(obj).reduce((runningSum, gvs) => {
    gvs.items.forEach((vs) => {
      runningSum += vs.concepts.length;
    });
    return runningSum;
  }, 0);
};
