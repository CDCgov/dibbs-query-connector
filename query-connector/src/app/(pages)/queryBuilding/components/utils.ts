import {
  DibbsValueSet,
  DibbsConceptType,
  Concept,
} from "@/app/shared/constants";
import { ConceptTypeToDibbsVsMap } from "@/app/utils/valueSetTranslation";
import { EMPTY_CONCEPT_TYPE } from "../utils";

export const CONDITION_DRAWER_SEARCH_PLACEHOLDER =
  "Search condition or category";
export const VALUESET_SELECTION_SEARCH_PLACEHOLDER =
  "Search labs, medications, conditions";
export const VALUESET_DRAWER_SEARCH_PLACEHOLDER = "Search by code or name";

type ValueSetWithRender = {
  render: boolean;
  concepts: FilterableConcept[];
};

export type FilterableConcept = Concept & {
  render: boolean;
};

export type FilterableValueSet = Omit<DibbsValueSet, "concepts"> &
  ValueSetWithRender;

/**
 * Helper function for search to filter out valuesets against a search param
 * @param searchFilter - search string
 * @param selectedValueSet - the active valueset displayed in the drawer
 * @param matchOnValueSetName - optional boolean to match the search filter on
 * the valueset name
 * @returns - a transformed list of concepts to display
 */
export function filterConcepts(
  searchFilter: string,
  selectedValueSet: FilterableValueSet | DibbsValueSet,
  matchOnValueSetName = true,
) {
  const casedSearchFilter = searchFilter.toLocaleLowerCase();
  const valueSetNameMatch =
    matchOnValueSetName &&
    selectedValueSet.valueSetName
      .toLocaleLowerCase()
      .includes(casedSearchFilter);

  const newConcepts = structuredClone(selectedValueSet.concepts);
  return newConcepts.map((concept) => {
    // if render has been set to false by a previous filter action (ie at the
    // valueset level, skip this concept)
    if (
      !matchOnValueSetName &&
      "render" in concept &&
      concept.render === false
    ) {
      return concept;
    }

    let toRender =
      valueSetNameMatch ||
      concept.code.toLocaleLowerCase().includes(casedSearchFilter) ||
      concept.display.toLocaleLowerCase().includes(casedSearchFilter);

    return { ...concept, render: toRender };
  });
}

/**
 * A helper function to filter a v
 * @param searchFilter - the search string to filter against
 * @param vs - the valueset to filter
 * @returns a valueset with the appropriate render flags set for itself/its concepts
 */
export function filterValueSet(searchFilter: string, vs: DibbsValueSet) {
  let vsNameMatch = vs.valueSetName.toLocaleLowerCase().includes(searchFilter);
  const conceptMatches = filterConcepts(searchFilter, vs);
  const curValueSet = {
    ...vs,
    render:
      // render a valueset if there's a match on the name or if there's a match
      // on any of its concepts
      vsNameMatch || conceptMatches.map((c) => c.render).some(Boolean),
  } as FilterableValueSet;
  curValueSet.concepts = conceptMatches;
  return curValueSet;
}

/**
 * Function to filter valuesets at the lab/meds/condition level
 * @param vsTypeLevelOptions - option to be filtered
 * @param searchFilter - search string
 * @returns a map of labs / conditions / meds to filtered concept types
 */
export function filterVsTypeOptions(
  vsTypeLevelOptions: ConceptTypeToDibbsVsMap,
  searchFilter: string,
) {
  const casedSearchFilter = searchFilter.toLocaleLowerCase();
  const filteredValueSets: {
    [conceptType in DibbsConceptType]: {
      [vsId: string]: FilterableValueSet;
    };
  } = structuredClone(EMPTY_CONCEPT_TYPE);

  Object.entries(vsTypeLevelOptions).forEach(([vsType, vsDict]) => {
    const curVsType = vsType as DibbsConceptType;
    Object.entries(vsDict).forEach(([vsId, vs]) => {
      // initialize the filterable concepts and value sets
      const curValueSet = filterValueSet(casedSearchFilter, vs);
      filteredValueSets[curVsType][vsId] = curValueSet;
    });
  });

  return filteredValueSets;
}
