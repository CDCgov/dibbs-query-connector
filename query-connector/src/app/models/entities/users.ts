import { QueryTableResult } from "@/app/(pages)/queryBuilding/utils";

export enum UserRole {
  ADMIN = "Admin",
  STANDARD = "Standard",
  SUPER_ADMIN = "Super Admin",
}
export interface User {
  id: string;
  username: string;
  first_name: string;
  last_name: string;
  qc_role: UserRole;
  userGroupMemberships?: UserGroupMembership[];
}

export interface UserGroup {
  id: string;
  name: string;
  member_size: number;
  query_size: number;
  members?: User[];
  queries?: QueryTableResult[];
}

export interface UserGroupMembership {
  membership_id: string;
  usergroup_name: string;
  usergroup_id: string;
  is_member: boolean;
}

export interface Query {
  id: string;
  name: string;
  userGroupMemberships?: UserGroupMembership[];
}
