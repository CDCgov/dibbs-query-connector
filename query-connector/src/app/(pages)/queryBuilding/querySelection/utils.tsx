import { ModalRef } from "@/app/ui/designSystem/modal/Modal";
import { DeleteModal } from "@/app/ui/designSystem/modal/deleteModal";
import { RefObject } from "react";
import { showToastConfirmation } from "@/app/ui/designSystem/toast/Toast";
import { DataContextValue } from "@/app/shared/DataProvider";
import { deleteQueryById } from "@/app/backend/query-building";
import { CustomUserQuery } from "@/app/models/entities/query";
import { UserGroupMembership } from "@/app/models/entities/users";

/**
 * Handles deleting a user query.
 * @param queryName - The name of the query to delete.
 * @param queryId - The ID of the query to delete.
 * @param queries - The current list of user queries.
 * @param setQueries - Function to update the state of queries.
 * @param context - The data context used to update shared state.
 */
export const handleDelete = async (
  queryName: string | undefined,
  queryId: string | undefined,
  queries: CustomUserQuery[],
  setQueries: React.Dispatch<React.SetStateAction<CustomUserQuery[]>>,
  context: DataContextValue | undefined,
) => {
  if (queryId) {
    const result = await deleteQueryById(queryId);
    if (result.success) {
      showToastConfirmation({
        body: `${queryName} has been deleted.`,
        variant: "success",
        duration: 2000,
      });
      const updatedQueries = queries.filter(
        (query) => query.query_id !== queryId,
      );
      setQueries(updatedQueries);

      if (context) {
        context.setData(updatedQueries);
      }
    } else {
      console.error(result.error);
    }
  } else {
    showToastConfirmation({
      heading: `Something went wrong`,
      body: `${queryName} couldn't be deleted. Please try again or contact us if the error persists`,
      variant: "error",
      duration: 2000,
    });
  }
};

/**
 * Confirms the deletion of a user query by toggling a modal.
 * @param queryName - The name of the query to confirm deletion for.
 * @param queryId - The ID of the query to confirm deletion for.
 * @param setSelectedQuery - Function to set the currently selected query for deletion.
 * @param modalRef - Reference to the modal component.
 */
export const confirmDelete = (
  queryName: string,
  queryId: string,
  setSelectedQuery: React.Dispatch<React.SetStateAction<SelectedQueryDetails>>,
  modalRef: RefObject<ModalRef>,
) => {
  setSelectedQuery({ queryName, queryId });
  modalRef.current?.toggleModal();
};

/**
 * Copies the query ID to the clipboard.
 * @param queryName - The name of the query to copy the ID for.
 * @param queryId - The ID of the query to copy.
 */
export const handleCopy = (queryName: string, queryId: string) => {
  navigator.clipboard
    .writeText(queryId)
    .then(() => {
      showToastConfirmation({
        body: `${queryName} ID copied successfully!`,
        duration: 2000,
      });
    })
    .catch((error) => {
      console.error("Failed to copy text:", error);
    });
};

/**
 *  Renders a modal to confirm the deletion of a user query.
 * @param modalRef - Reference to the modal component.
 * @param selectedQuery - The currently selected query for deletion.
 * @param handleDelete - Function to handle the deletion workflow.
 * clean up the internal state after deletion.
 * @param group - The current user's group.
 * @param queries - The current list of user queries.
 * @param setQueries - Function to update the state of queries.
 * @param context - The data context used to update shared state.
 * @param setSelectedQuery - Function to update the currently selected query to
 * @returns The JSX element for the modal.
 */
export const renderModal = (
  modalRef: RefObject<ModalRef>,
  selectedQuery: SelectedQueryDetails | null,
  handleDelete: (
    queryName: string | undefined,
    queryId: string | undefined,
    queries: CustomUserQuery[],
    setQueries: React.Dispatch<React.SetStateAction<CustomUserQuery[]>>,
    context: DataContextValue,
  ) => void,
  group: CustomUserQuery["userGroupMemberships"],
  queries: CustomUserQuery[],
  setQueries: React.Dispatch<React.SetStateAction<CustomUserQuery[]>>,
  context: DataContextValue,
  setSelectedQuery: React.Dispatch<React.SetStateAction<SelectedQueryDetails>>,
): JSX.Element => {
  return (
    <DeleteModal
      modalRef={modalRef}
      heading="Confirm Deletion"
      description={`Are you sure you want to delete "${
        selectedQuery ? selectedQuery.queryName : ""
      }"? This action cannot be undone.`}
      onDelete={() => {
        if (selectedQuery) {
          handleDelete(
            selectedQuery.queryName,
            selectedQuery.queryId,
            queries,
            setQueries,
            context,
          );

          setSelectedQuery({ queryName: undefined, queryId: undefined });
        }
      }}
    />
  );
};

export type SelectedQueryDetails = {
  queryName?: string;
  queryId?: string;
};
export type SelectedQueryState = SelectedQueryDetails | null;
