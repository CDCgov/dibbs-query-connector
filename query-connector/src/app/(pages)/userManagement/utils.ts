import { useSession } from "next-auth/react";
import { useContext } from "react";
import { DataContext } from "@/app/shared/DataProvider";
import { isAuthDisabledClientCheck } from "@/app/utils/auth";
import { UserRole } from "@/app/models/entities/users";

/**
 * Function that retrieves the role value from the current session object
 * @returns - The UserRole for the current logged-in user
 */
export function getRole() {
  const { data: session } = useSession();
  const ctx = useContext(DataContext);
  const isAuthDisabled = isAuthDisabledClientCheck(ctx?.runtimeConfig);

  return isAuthDisabled ? UserRole.SUPER_ADMIN : session?.user.role;
}

export const RoleDescriptons = {
  "Super Admin":
    "Manage user permissions; create and manage user groups; view audit logs; configure FHIR servers; create, assign, and run queries",
  Admin: "Create, assign, and run queries",
  Standard: "Only run queries",
};

export type UserManagementMode =
  | "create-user"
  | "edit-user"
  | "remove-user"
  | "select-groups"
  | "create-group"
  | "add-members"
  | "add-queries"
  | "remove-group"
  | "closed";

type ModalMap = {
  [mode in UserManagementMode]: {
    heading: string;
    buttonText: string;
    secondaryBtnText: string;
    prevStep?: UserManagementMode;
  };
};

export const ModalStates: ModalMap = {
  "create-user": {
    heading: "New user",
    buttonText: "Next: Add to user groups",
    secondaryBtnText: "",
    prevStep: "closed",
  },
  "edit-user": {
    heading: "Edit user",
    buttonText: "Next: Edit user groups",
    secondaryBtnText: "",
    prevStep: "closed",
  },
  "select-groups": {
    heading: "Add to user groups",
    buttonText: "Add User",
    secondaryBtnText: "Back",
    prevStep: "create-user",
  },
  "create-group": {
    heading: "Create user group",
    buttonText: "Next: Assign members",
    secondaryBtnText: "Cancel",
    prevStep: "closed",
  },
  "add-members": {
    heading: "Add members",
    buttonText: "",
    secondaryBtnText: "Next: Add to user groups",
    prevStep: "create-group",
  },
  "add-queries": {
    heading: "TK",
    buttonText: "TK",
    secondaryBtnText: "Next: Add to user groups",
    prevStep: "add-members",
  },
  "remove-user": {
    heading: "TK",
    buttonText: "TK",
    secondaryBtnText: "Next: Add to user groups",
    prevStep: "closed",
  },
  "remove-group": {
    heading: "TK",
    buttonText: "TK",
    secondaryBtnText: "Next: Add to user groups",
    prevStep: "closed",
  },
  closed: {
    heading: "",
    buttonText: "",
    secondaryBtnText: "Next: Add to user groups",
    prevStep: "closed",
  },
};
