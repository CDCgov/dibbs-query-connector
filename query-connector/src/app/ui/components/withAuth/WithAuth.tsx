"use client";

import React, { useContext } from "react";
import { UserRole } from "@/app/models/entities/users";
import { useSession } from "next-auth/react";
import { redirect, usePathname } from "next/navigation";
import { pagesConfig } from "@/app/shared/page-routes";
import { DataContext } from "@/app/shared/DataProvider";
import { isAuthDisabledClientCheck } from "@/app/utils/auth";

/**
 * @param root0 AuthPageGuard component props
 * @param root0.children Content protected behind the auth guard
 * @returns A wrapper component that checks if the user is logged in and their role before displaying content
 * Is a component uses withAuth but the page access has not being set then if will redirect to /unauthorized,
 * unless auth is disabled in which case the component will render without issues.
 */
const WithAuth: React.FC<React.PropsWithChildren> = ({ children }) => {
  const { status, data: session } = useSession();
  const role = session?.user?.role || "";
  const ctx = useContext(DataContext);
  const isAuthDisabled = isAuthDisabledClientCheck(ctx?.runtimeConfig);
  const path = usePathname();
  const access = pagesConfig[path]?.roleAccess ?? [];

  if (
    isAuthDisabled ||
    (status === "authenticated" && access.includes(role as UserRole))
  ) {
    return <>{children}</>;
  } else if (status === "loading") {
    return null;
  } else if (status === "unauthenticated") {
    redirect("/");
  } else {
    return (
      <div className="main-container__wide">
        <h1>Unauthorized Access</h1>
        <p>You are not authorized to access this page.</p>
      </div>
    );
  }
};

export default WithAuth;
