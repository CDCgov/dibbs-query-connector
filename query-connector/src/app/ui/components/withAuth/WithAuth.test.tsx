import { render, screen } from "@testing-library/react";
import WithAuth from "./WithAuth";
import * as nextAuthReact from "next-auth/react";
import * as authUtils from "@/app/utils/auth";
import { UserRole } from "@/app/models/entities/users";
import * as nextNavigation from "next/navigation";
import * as Utils from "@/app/(pages)/userManagement/utils";
import { Session } from "next-auth";
import { SessionContextValue } from "next-auth/react";

jest.mock("next-auth/react");

jest.mock("@/app/utils/auth", () => ({
  isAuthDisabledClientCheck: jest.fn().mockReturnValue(false),
}));

jest.mock("@/app/(pages)/userManagement/utils", () => ({
  getContextRole: jest.fn().mockReturnValue("Standard"),
}));

jest.mock("next/navigation", () => ({
  redirect: jest.fn(),
  usePathname: jest.fn(),
}));

describe("WithAuth component (page access guard)", () => {
  const DummyComponent = () => <>Special content</>;

  beforeAll(() => {
    jest
      .spyOn(nextNavigation, "usePathname")
      .mockReturnValue("/userManagement");
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it("Renders content when auth is disabled", () => {
    jest
      .spyOn(authUtils, "isAuthDisabledClientCheck")
      .mockReturnValueOnce(true);

    render(
      <WithAuth>
        <DummyComponent />
      </WithAuth>,
    );

    expect(screen.getByText("Special content")).toBeInTheDocument();
  });

  it("Renders content if user is logged in and has role access", () => {
    const session: SessionContextValue = {
      data: { user: { role: UserRole.SUPER_ADMIN } } as Session,
      status: "authenticated",
      update: jest.fn(),
    };
    jest.spyOn(Utils, "getContextRole").mockReturnValueOnce("Super Admin");
    jest.spyOn(nextAuthReact, "useSession").mockReturnValueOnce(session);

    render(
      <WithAuth>
        <DummyComponent />
      </WithAuth>,
    );
    expect(document.body).toMatchSnapshot();

    expect(screen.getByText("Special content")).toBeInTheDocument();
  });

  it("Renders unauthorized if user is logged in and does not have role access", () => {
    const session: SessionContextValue = {
      data: { user: { role: UserRole.STANDARD } } as Session,
      status: "authenticated",
      update: jest.fn(),
    };

    jest.spyOn(nextAuthReact, "useSession").mockReturnValueOnce(session);

    render(
      <WithAuth>
        <DummyComponent />
      </WithAuth>,
    );
    expect(
      screen.getByText("You are not authorized to access this page."),
    ).toBeInTheDocument();
  });
  it("Redirects user to landing page if user is not logged in", () => {
    const session: SessionContextValue = {
      data: null,
      status: "unauthenticated",
      update: jest.fn(),
    };

    jest.spyOn(nextAuthReact, "useSession").mockReturnValueOnce(session);
    const redirectSpy = jest.spyOn(nextNavigation, "redirect");

    render(
      <WithAuth>
        <DummyComponent />
      </WithAuth>,
    );
    expect(redirectSpy).toHaveBeenCalledWith("/");
  });
  it("Renders empty content while the session is being retrieved", () => {
    const session: SessionContextValue = {
      data: null,
      status: "loading",
      update: jest.fn(),
    };

    jest.spyOn(nextAuthReact, "useSession").mockReturnValueOnce(session);
    render(
      <WithAuth>
        <DummyComponent />
      </WithAuth>,
    );
    expect(screen.queryByText("Special content")).not.toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });
});
