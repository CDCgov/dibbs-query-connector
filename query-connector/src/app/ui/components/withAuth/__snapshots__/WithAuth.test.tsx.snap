// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithAuth component (page access guard) Renders empty content while the session is being retrieved 1`] = `
<body>
  <div />
</body>
`;

exports[`WithAuth component (page access guard) Renders unauthorized if user is logged in and does not have role access 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <h1>
        Unauthorized Access
      </h1>
      <p>
        You are not authorized to access this page.
      </p>
    </div>
  </div>
</body>
`;
